### PR TITLE
Patches for adaptive time stepping

### PR DESCRIPTION
**Pull request by wenqing**
This PR contains updates or bugfixing for the adaptive time stepping implementation:

- stop program when a rejected time step gets an unchanged  new step size if the previous step is also rejected.
- reset the initial guess of global x vector to its old value if a time step is rejected. Actually only the Picard  nonlinear solver needs  this computation.
- only display the accepted time step number as step number.

Besides, a test for the case of Richards flow with adaptive time stepping is also included in this PR.

### CODE REVIEW

**Comment by wenqing** [comment]
@waltherm The updates are here, and I am going to test your problem with Tom's branch.

File: NumLib/ODESolver/TimeDiscretization.h [hunk 1]
```diff
@@ -142,6 +142,17 @@ class TimeDiscretization
     virtual void pushState(const double t, GlobalVector const& x,
                            InternalMatrixStorage const& strg) = 0;
 
+    /*! Indicate that the current timestep is rejected and that the computation
+     *  of the present time step will be restarted with a new time step size.
+     *
+     * \param t    The current time step.
```
Comment (endJunction) [inline NumLib/ODESolver/TimeDiscretization.h]: t seems to be unused in all functions...

Comment (wenqing) [inline NumLib/ODESolver/TimeDiscretization.h]: Oh it is! Will be removed.

Comment (wenqing) [inline NumLib/ODESolver/TimeDiscretization.h]: removed `t` by amending the commit of ``[dt] Set current x to old x when a step is ...``.

**Comment by waltherm** [comment]
@wenqing thanks, let me know, if I can help and test something...

File: NumLib/ODESolver/TimeDiscretization.h [hunk 2]
```diff
@@ -142,6 +142,14 @@ class TimeDiscretization
     virtual void pushState(const double t, GlobalVector const& x,
                            InternalMatrixStorage const& strg) = 0;
 
+    /*! Indicate that the current timestep is rejected and that the computation
+     *  of the present time step will be restarted with a new time step size.
```
Comment (endJunction) [inline NumLib/ODESolver/TimeDiscretization.h]: Sounds a little misleading to me: The function does not indicate anything. Maybe:
> Restores previous time step result into the given vector.
> Used for repeating of the time step with new time step size when the current timestep is rejected.

**Comment by endJunction** [comment]
@waltherm Please have a look if this resolves the time stepping problems for your test cases. Then it can be merged.

**Comment by waltherm** [comment]
@endJunction will do and report.

**Comment by wenqing** [comment]
@waltherm  By took Tom's branch and rebased on this PR, I just tested your example of RichardsComponentTransportProcess with the following time stepping input
```
                 <time_stepping>
                     <type>EvolutionaryPIDcontroller</type>
                     <t_initial> 0.0 </t_initial>
                     <t_end> 20 </t_end>
                     <dt_guess> 0.1 </dt_guess>
                     <dt_min> 0.01 </dt_min>
                     <dt_max> 5 </dt_max>
                     <rel_dt_min> 0.01 </rel_dt_min>
                     <rel_dt_max> 5 </rel_dt_max>
                     <tol> 0.2 </tol>
                 </time_stepping>
```
You can also use PerComponentDeltaX for convergence criterion.

**Comment by waltherm** [comment]
Thanks, @wenqing for the patches. Here are some comments.

Issues **solved**:
- ogs stops now, if the time step is repeated, but the minimum time step size was used.
- Warning message for time step repetition now correctly refers to current time step number.

Issues **open**:
- There is no additional information why a time step was rejected, although the iteration converged.
- Repetitions of a time step are not marked with a repetition index (not really important).

Wording **suggestion**:
- "warning: Time step 1 is rejected. The computation is back to the previous time." 
-> "warning: Time step 1 was rejected and will be repeated with a reduced time step size."
- "error: 	This time step is rejected and the new computed step size is the same as that was just used. Suggest to adjust the parameters of the time stepper or try other time stepper." 
-> "error: Minimum time step size was rejected. Simulation is halted."

**General**: 
- I tried to run a simulation with RichardsComponentTransport and varied the maximum time step size. The time step algorithm will increase the time step to really large values, although the simulation shows clear temporal and spatial oscillations of the primary variable. Practically, this makes automatic time stepping not very useful. How can we solve this? Here are figures for the oscillations: 
spatial_osc
temp_osc

File: NumLib/ODESolver/TimeDiscretization.h (hunk 2 above)
Comment (wenqing) [inline NumLib/ODESolver/TimeDiscretization.h]: Changed.

**Comment by wenqing** [comment]
@waltherm Added a commit about what ``Issues open `` and  ``Wording suggestion``in you comment.
The increase ratio of step size can be changed by setting the tag of ``<rel_dt_max>  </rel_dt_max>``.

**Comment by waltherm** [comment]
Just a minor issue: output for repeatedly rejected time steps is now always: "warning: Time step 1 was rejected 1 times and it will be repeated with a reduced step size.", it never increases to "2 times or more".

**Pull request merged**